File: backend/app/schemas/replay.py
"""Replay timeline schemas — ordered drawing ops, not canvas snapshots.

Events mirror committed canvas mutations (never ephemeral previews).
Payloads reuse WhiteboardShape / HistoryOp shapes from ``schemas.canvas``.
"""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


ReplayEventType = Literal[
    "shape.created",
    "shape.updated",
    "shape.deleted",
    "canvas.cleared",
    "undo",
    "redo",
]

ReplayTool = Literal[
    "pencil",
    "bezier",
    "rectangle",
    "ellipse",
    "arrow",
    "fill",
    "eraser",
]


class ReplayEvent(BaseModel):
    """One deterministic drawing operation on the timeline."""

    t: int = Field(ge=0, description="Milliseconds since round start")
    type: ReplayEventType
    player_id: str = Field(min_length=1, max_length=64)
    tool: ReplayTool | None = None
    payload: dict[str, Any] = Field(default_factory=dict)

    @field_validator("player_id", mode="before")
    @classmethod
    def coerce_player_id(cls, value: Any) -> str:
        return str(value)


class DrawingReplayMeta(BaseModel):
    drawing_id: UUID
    session_id: UUID
    turn_number: int
    word: str
    author_id: UUID
    author_name: str | None = None
    duration_ms: int = Field(ge=0)
    event_count: int = Field(ge=0)
    published_at: str | None = None


class DrawingReplay(BaseModel):
    """Single-round replay: metadata + ordered events."""

    version: Literal[1] = 1
    meta: DrawingReplayMeta
    events: list[ReplayEvent]


class GameReplayResponse(BaseModel):
    """All published round replays for a finished (or in-progress) game session.

    Extensible: extra fields can be added without breaking clients that ignore them.
    """

    version: Literal[1] = 1
    game_id: UUID = Field(description="Game session id")
    room_id: UUID | None = None
    drawings: list[DrawingReplay]


class DrawingReplayResponse(DrawingReplay):
    """Alias response model for GET /drawings/{id}/replay."""

    pass

File: frontend/services/replay.ts
/**
 * HTTP client for drawing / game replay timelines.
 */

import { apiFetch } from "@/lib/api";
import type {
  DrawingReplay,
  GameReplay,
  ReplayEvent,
} from "@/features/replay/types";

interface DrawingReplayApi {
  version: 1;
  meta: {
    drawing_id: string;
    session_id: string;
    turn_number: number;
    word: string;
    author_id: string;
    author_name: string | null;
    duration_ms: number;
    event_count: number;
    published_at: string | null;
  };
  events: ReplayEvent[];
}

interface GameReplayApi {
  version: 1;
  game_id: string;
  room_id: string | null;
  drawings: DrawingReplayApi[];
}

function mapDrawing(raw: DrawingReplayApi): DrawingReplay {
  return {
    version: 1,
    meta: {
      drawing_id: raw.meta.drawing_id,
      session_id: raw.meta.session_id,
      turn_number: raw.meta.turn_number,
      word: raw.meta.word,
      author_id: raw.meta.author_id,
      author_name: raw.meta.author_name,
      duration_ms: raw.meta.duration_ms,
      event_count: raw.meta.event_count,
      published_at: raw.meta.published_at,
    },
    events: Array.isArray(raw.events) ? raw.events : [],
  };
}

export async function fetchGameReplay(gameId: string): Promise<GameReplay> {
  const data = await apiFetch<GameReplayApi>(`/games/${gameId}/replay`);
  return {
    version: 1,
    game_id: data.game_id,
    room_id: data.room_id,
    drawings: (data.drawings ?? []).map(mapDrawing),
  };
}

export async function fetchDrawingReplay(
  drawingId: string,
): Promise<DrawingReplay> {
  const data = await apiFetch<DrawingReplayApi>(
    `/drawings/${drawingId}/replay`,
  );
  return mapDrawing(data);
}
